### ISSUE TITLE
Open in Coder

### ISSUE THREAD

**Issue by bpmct** [coder/coder#3981]
Some button that links to the template create page for a given template that people can put in a README :)

<img width="891" alt="Screen Shot 2022-09-12 at 10 06 22 AM" src="https://user-images.githubusercontent.com/22407953/189689433-694ffaea-fbe2-4e73-9e2c-b5b396238037.png">

And perhaps some way for people to generate these "sharable" buttons. In Coder Classic, we [had it in the images page](https://i.imgur.com/z4K2Yl4.png). Perhaps as a tab in the template page?

https://railway.app/button is pretty cool too :)

**Comment by sharkymark** [comment]
This is a popular feature with Coder v1 customers.

Why?

Devs are familiar with their Git provider as part of their daily flow. So they can jump to Coder directly from a repo they need to work on. (which takes you to a proper template)

**Comment by bpmct** [comment]
Curious what y’all think of the following behavior:

- Open in Coder button is often embedded inside repos or internal wikis for specific projects/repos. Often a template can support multiple projects (for example, our dogfood template can be used to develop `v1`, `v2`, and `code-server`). 
- Open in Coder button leads to the  template or workspace create page (@jsjoeio shared this idea via Slack)
- Open in Coder button can pre-fill (and potentially grey out) some parameters based on URL query parameters
- The template requires the user to auth with their Git Provider so an automatic clone can occur on behalf of the user #3078
- @ghuntley shared that one benefit a feature like this not having to do `git clone`: The template admin can define clone logic in the template/parameters/[startup_script](https://registry.terraform.io/providers/coder/coder/latest/docs/resources/agent#startup_script) so the cloning mechanism isn’t limited to the "open in coder" button and is also not a black box. (more below)

<img width="400" src="https://user-images.githubusercontent.com/22407953/189688062-ba974f8f-0424-498b-b8dd-b3a33715fbac.png" />

```hcl
# clone logic in the template instead of hardcoded.
resource "coder_agent" "dev" {
  ...
  startup_script =<<EOF
  # Check if the project directory already exists 
  if [ ! -d "project"]; then
    echo "project" is empty... Cloning"
    git clone ${var.git_repo}
  fi
  EOF
}
 ```

## Authentication

Right now, there’s a problem: if the user hasn’t authenticated their workspace with their git provider (token or SSH), any git clone steps would fail since the user doesn’t have permission. If a user is coming from a repo and it is their first workspace, it is highly likely they never added their keys to GitHub, or perhaps they are required to auth via token (we have v1 customers and v2 prospects who do this). One solution is to add a git provider mechanism into Coder where users are prompted to link before creating a workspace. We’ve had 2 prospects request this feature (#3078)

## Open in Coder

The auto-cloning in v1 is also a bit of a black box. Like the entrypoint, code-server version, and projector version, it’s hardcoded in v1 as a part of the startup logic. This also leads to 2 different flows (not mentioning v1 workspace templates). I tried to make a basic diagram (and spelled parameter wrong)!

![Open in Coder (classic)](https://user-images.githubusercontent.com/22407953/189684929-0914bca8-a3ff-4dd4-ae22-66043fdb6dd3.png)

**Comment by bpmct** [comment]
@jsjoeio I love the simplicity! I'm assuming this means you consider the "auto-clone" functionality less important for the feature or up to the template to implement? For monorepos or 1-repo-per-template use cases, this makes sense to me, but not sure how this would work if 1 template is used across multiple teams.

Perhaps something we could consider down the road.

**Comment by jsjoeio** [comment]
Spoke to Ben offline about this. I'm going to create a mini-design spec (instead of full-blown RFC) and will post here before starting any work.

**Comment by jsjoeio** [comment]
@bpmct would love your thoughts and feedback on this mini-design spec. Also happy to add diagrams if needed. I will wait until I get your 👍🏼 

---

## Problem
Design an “Open in Coder” button.

## Requirements
As an Admin, I want to add an “Open in Coder” button to my repository so my users can navigate to their workspace faster.
- I also want to prefill workspace form values.

As a User, I want to navigate to my Workspace by clicking an “Open in Coder” button which
- I also want my workspace to start if it’s currently stopped.

## Assumptions & Tradeoffs
To ensure this scales, we are decoupling the “clone repo” logic from this workflow. We assume Admins will handle any repo cloning directly in their templates. We assume this workflow is tied to templates.

## API
This feature will be implemented via the following format:  <deployment-url> + <optional query parameters>. 

### deployment url
Where their deployment lives i.e. https://dev.coder.com/ This is the bare minimum needed to use the “Open in Coder” button.

### optional query parameters
All query parameters are optional. The following can be used

- `template`
  - Name of a valid Template
  - Type: `string`
  - Default: “”
  - If Template doesn’t exist, navigate User to Templates page and display error.
- `createWorkspace`
  - Type: `boolean`
  - Default: `true`
  - If User already has Workspace created with Template, navigate to Workspace page and start Workspace.
    - If User has multiple Workspaces with Template and navigate to most recently used Workspace page.
  - If User does not have Workspace created with Template, navigate User to Workspace form.
- `autofill`
  - Type: `boolean`
  - Default: `true`
  - If User does not have Workspace created with Template, autogenerate values, create Workspace and navigate User to Workspace page.
- `autostart`
  - Type: `boolean`
  - Default: `true`
  - If a User already has a Workspace created with Template, it will autostart if currently stopped.

**Comment by bpmct** [comment]
Hey @jsjoeio - looks solid. A few clarifying questions:

> As an Admin, I want to add an “Open in Coder” button to my repository so my users can navigate to their workspace faster.

Where would an admin go to generate the "Open in Coder" button?

> I also want to prefill workspace form values.
> If User does not have Workspace created with Template, autogenerate values, create Workspace and navigate User to Workspace page.

What kind of values will be autogenerated? For example, with this AWS template, would it pick random parameter values? With text-based parameters (e.g. docker image) would it be blank?

<img width="300" alt="Screen Shot 2022-09-14 at 2 50 40 PM" src="https://user-images.githubusercontent.com/22407953/190249003-99fb01a6-5cc1-4f6b-86d5-c0c4fd923fc7.png">

> `autostart`

Thoughts on changing this to `start`? This slightly confused me with our feature that automatically starts workspaces at a given time every day.

> template

I saw the `template` field is optional. Where would the link go if a template is not specified? 

> createWorkspace

Thoughts on doing 3 values here? `always` `first_of_type` and `never` We have some users who use a single template (image in Coder classic) across multiple workspaces. One for each project/client/repo, for example.

**Comment by jsjoeio** [comment]
Great questions!

> Where would an admin go to generate the "Open in Coder" button?

For first version, they would read about it in the docs and see an example like 

"Add an 'Open in Coder' button to your `README.md` using this format: `[![open in coder button](https://coder.com/assets/button.png)](<deployment-url>&parameters)`.

After first version, we could add a form that generates it for them into the Admin Dashboard (or wherever you see fit).

> I can imagine a workspace name could be random (h3DdjNG) but not sure how this would work with template parameters.
> For example, with this AWS template, would it pick random parameter values? With text-based parameters (e.g. docker image) would it be blank?

Ah, that's a great point...Maybe we punt the autofill feature for now?

> Thoughts on changing this to start? This slightly confused me with our feature that automatically starts workspaces at a given time every day.

Good point. Let's go with `startWorkspace` to be extra clear.

> Thoughts on doing 3 values here? always first_of_type and never

Good idea! Let's change `first_of_type_` to `firstOfType` since I think that's a bit easier to type.

**Comment by bpmct** [comment]
> For first version, they would read about it in the docs and see an example like
> "Add an 'Open in Coder' button to your README.md using this format: `[![open in coder button](https://coder.com/assets/button.png)](<deployment-url>&parameters)`.
> After first version, we could add a form that generates it for them into the Admin Dashboard (or wherever you see fit).

I like the emphasis on iterative shipping but think the generation UI is core to the experience (e.g. not calling this issue "done" until that is added). @kylecarbs showed me an interesting Figma concept with tabs on the "templates page" where there is a tab to generate the embed button. I could give that a shot if you'd prefer to just do the backend/routing

>  Ah, that's a great point...Maybe we punt the autofill feature for now?

What if you could specify parameter values via the URL? For example `<deployment-url>?template=idk&region=us-west-2&git-repo=https://github.com`? Kind of like how Heroku has [paramaterizing buttons](https://devcenter.heroku.com/articles/heroku-button#parametrizing-buttons) which pre-fill the input

> firstOfType

👍🏼

**Comment by jsjoeio** [comment]
> I like the emphasis on iterative shipping but think the generation UI is core to the experience (e.g. not calling this issue "done" until that is added). @kylecarbs showed me an interesting Figma concept with tabs on the "templates page" where there is a tab to generate the embed button. 

That make sense 👍🏼 

> I could give that a shot if you'd prefer to just do the backend/routing

Sure! I'll do all the work stated above and you focus on the form/experience for authoring the button. Do you mind doing the design for the button too? 

> What if you could specify parameter values via the URL? For example <deployment-url>?template=idk&region=us-west-2&git-repo=https://github.com/org/project1? Kind of like how Heroku has [paramaterizing buttons](https://devcenter.heroku.com/articles/heroku-button#parametrizing-buttons) which pre-fill the input

Great idea! That shouldn't be too difficult to support. I think we'll still take an iterative approach and I'll do that part last.

If the rest sounds good, am I clear to begin working on this? 👍🏼

**Comment by bpmct** [comment]
Let's do it!

**Comment by jsjoeio** [comment]
Per feedback in https://github.com/coder/coder/pull/4124, we decided it makes more sense to use the `<deployment-url>/templates<template-name>` as the base URL.

**Comment by jsjoeio** [comment]
@bpmct @kylecarbs planning to start on Phase 2 today: `createWorkspace` param. I want to get the 👍🏼 before starting.

Here's what I'm thinking:

`createWorkspace` query param (added to the /templates/<template-name> url)

- optional: doesn't have to be used
- valid options: `always`, `firstOfType`

`always` - always redirect to Create Workspace page for template.

`firstOfType` - only redirect to Create Workspace page for template if user doesn't already have a workspace made from template.

Feedback welcome!

**Comment by jsjoeio** [comment]
Spoke to @kylecarbs and @bpmct offline about this. We're going to go the simple route for now. Ben can share more.

**Comment by bpmct** [comment]
TLDR is we're going to simply provide a button for now and evaluate pre-filled params down the road.

**Comment by MrPeacockNLB** [comment]
I'am really looking forward for this feature

**Comment by bpmct** [comment]
Also consider a "personalize" step

**Comment by bpmct** [comment]
How this will work (still subject to change, however):

1. A template admin edits the template to specify what repo(s) it is compatible with (or regex-like `https://github.com/*`, `https://gitlab.company.com/myorg/*`).
2. A template admin can embed an "Open in Coder" button on their repo or internal wiki
3. When a user clicks the button in the repo, they can pick a compatible template and then create/use an existing workspace
4. The workspace will attempt to clone the repo, with logic editable via the Coder agent's startup script. If the user has not linked their git provider yet, Coder will ask them to connect before they can enter their workspace.

Users can also create a workspace from a repo via the Coder dashboard:

https://user-images.githubusercontent.com/22407953/213309540-325bc8f6-cde4-4511-aa51-ff9c4fba5df7.mov

**Comment by ehfd** [comment]
https://gitlab.com/gitlab-org/gitlab/-/issues/384217

![image](https://user-images.githubusercontent.com/8457324/235712940-3d5ea42b-ed5e-43d2-9757-53cf27842322.png)

Would it be possible to contribute this to GitLab so that users may use Coder without adding the button in the markdown file?
